### PR TITLE
 WIP Qt5 completion in feature/cmake

### PR DESCRIPTION
**Pull request by eleaver**
1. This will be Work In Progress pending CR, Windows build, tests, and dependencies source upload deposition.

2. Features:
   a. Target platforms CentOS-7, CentOS-8, and Windows
   b. Qt5 and Qwt6 mostly provided by Trevor Clarke
       Builds with system Qt-5.9.7 on CentOS-7 and Qt-5.12.5 on CentOS-8. (There are other options.)
   c. Nvidia's binary Cg has been replaced by OpenGL glsl code provided by Jonathan Overholser.
   d. Most Opticks Dependencies are now provided either by system yum/dnf packages, the few that are not may be
       compiled from source. The opticks/Dependencies folder and its downloads need not be used by this branch,
       although it's layout may be duplicated for the remaining non-system dependencies. 
       * Base requirements are:         ehs  ffmpeg  opencollada  ossim  pme  raptor yaml-cpp
       * CentOS-7  needs base plus tbb.
       * CentOS-8  needs base plus xqilla. (Its system tbb is fine).
       * Fedora-32 needs base plus opencv/opencv2 3.4.10 or so, as its system opencv 4.2 is too new. Fedora may
                  also use it's system raptor package.
       * Source urls are given in the attached README.dependencies.
   e. The above Dependency changes entail some version bumps, with conditional changes to some APIs. Not all
       dependency packages have a version.h header consistent between versions; in such cases version
       numbers are determined by the relevant cmake/Modules/Find<Package>.cmake. 
       i.   yaml-cpp version 0.3.0
            * There have been some minor API changes since 0.2.5, these are enabled via conditional compilation.
            * Version number is determined in FindYamlCpp.cmake
       ii.  ossim version 1.9.0
            * ossim 1.9.0 does not have a SVDSolve() method; I've substituted Eigen's but it needs testing.
            * a new PluginUtilities/Interfaces/OssimVersion.h file aids version determination. 
       iii. ffmpeg version 0.7.17 brings
           avcodec    52.123.0
           avdevice   52.5.0
           avfilter    1.80.0
           avformat   52.11.0
           avutil     50.43.0
           avpostproc 51.2.0
           swscale     0.14.1
           * ffmpeg no longer has img_convert() function, it has been
             replaced in Opticks' MovieExporter plugin by sws_scale().
           * ffmpeg APIs are traditionally unstable; hoipefully this
             newer 0.7.17 version will remain readily available.

3. CentOS-7:
    a. CentOS-7's system gcc-4.8.5 is too old to use with its Qt-5.9.7 system package, therefore
        i.   yum install centos-release-scl      and enable /etc/yum.repos.d/CentOS-SCLo-scl.repo
        ii.  yum install scl-utils
        iii. yum install devtoolset-8-*
            This enables a gcc-8.3.1 build environment. (CentOS-8's system compiler is also gcc-8.3.1)

4. Qt5:
     a. With Qt5, QFile no longer reliably syncs its file to disk. If you want to write a file to disk,
          you should use QSaveFile instead. I've changed Opticks' journal and log files to QSaveFiles,
          but Opticks uses a *lot* of QFiles, and every one of them will need be checked to see if they
          are written. See Utilities/MessageLogImp and MessageLogMgrImp for example usages of QSaveFile.
      b. Edit 9/18/2020: I'm hoping the problem is that the Journal and Log files both set their QFile* pointers to
           new QTemporaryFile(filename) rather than new new QFile(filename). I've replaced those two cases
           with new QSaveFile(filename), but will do further testing to see if new QFile(filename) will also work.
       c. Edit 9/18/2020: Such appears to be the case. I've tried replacing QTemporaryFile with QFile for mpJournal
           and mpLogFile, and as QFiles the mpJournal and mpLogFile do indeed flushed to disk. I've retained them
           as QSaveFiles for now, but the rest of Opticks' QFiles are quite likely okay.

5. Qwt6:
   a. dnf install qwt-qt5-devel on CentOS-8 (or Fedora32). The qwt-qt5 package provides Qwt6 for system Qt5.
   b. There is no qwt-qt5 package on CentOS-7.  So compile qwt-6.1.4 (or newer) from source,
      either from qwt-6.1.4.zip or svn, and install in for example /usr/local/qwt-6.1.4-qt-5.9.7
      i.   unzip qwt-6.1.4.zip
      ii.  cd qwt-6.1.r and edit qwtconfig.pri to set your desired QWT_INSTALL_PREFIX
           (I like to include the QT_VERSION in my install location path.)
      iii. "scl enable devtoolset-8 bash" to get the dts-8 bash prompt
      iv.  /usr/bin/cmake-qt5 -r; make; make install
      v.   Set this Qwt install path in the aforementioned ConfigureOpticks.sh

Note: Fedora is not an Opticks supported OS; it's where I do my development. 
      CentOS-7 is Opticks' official Linux platform, with CentOS-8 rsn.

Attachments:
ConfigureOpticks.sh
README.dependencies
DependencyConfigs.zip

[ConfigureOpticks.sh.txt](https://github.com/opticks-org/opticks/files/5243117/ConfigureOpticks.sh.txt)
[README.dependencies.txt](https://github.com/opticks-org/opticks/files/5243123/README.dependencies.txt)
[DependencyConfigs.zip](https://github.com/opticks-org/opticks/files/5243125/DependencyConfigs.zip)

### CODE REVIEW

**Comment by tclarke** [comment]
I'll try and get around to reviewing this today but it might not be until next week.

**Comment by eleaver** [comment]
I've a preliminary Spectral Qt5 port. It's plugins load on CentOS-7, CentOS-8, and Fedora 32, but it's CMake tree needs some cleanup.

**Comment by tclarke** [comment]
Could you merge in the lastest master to this branch? I would like to get this work into master and need to resolve the qt5 changes that went into master a while ago.

File: Code/application/PlugInUtilities/MatrixFunctions.cpp [hunk 1]
```diff
@@ -364,3 +366,67 @@ bool MatrixFunctions::solveLinearEquation(double* pResult, const double** pLhs,
    memcpy(pResult, pData, resultVector.Storage() * sizeof(pData[0]));
    return true;
 }
+#else
+// ossim 1.9.0 doesn't have an explicit SVDSolve() function, so we'll use Eigen's
```
Comment (tclarke) [inline Code/application/PlugInUtilities/MatrixFunctions.cpp:370]: There are a number of changes that were made to ossim that are not in their trunk (opticks runs off a fork) which is why there is no SVDSolve() available in stock 1.9.0. I would like to get the ossim dependency out of the required build and limit it to builds requiring NITF. This is a good first step (eigen wasn't available when this code was first written).

If you see any other changes like this, they can go into master as well. Once I have most things up to date and working with clean dependencies I intend to go over all the other ossim changes and either remove the need for them or push the changes to the ossim project.

File: Code/application/PlugIns/src/DataFusion/TiePointPage.cpp [hunk 2]
```diff
@@ -294,7 +294,7 @@ void TiePointPage::gcpLayerDeleted(Subject &subject, const string &signal, const
          GcpLayer* pCopy = NULL;
          if (pView != NULL)
          {
-            pCopy = static_cast<GcpLayer*>(pGcpLayer->copy(false));
+            pCopy = static_cast<GcpLayer*>(pGcpLayer->copy(0));
```
Comment (tclarke) [inline Code/application/PlugIns/src/DataFusion/TiePointPage.cpp:297]: why were these changed to 0 from false?

File: Code/application/Utilities/DataVariantFactoryImp.cpp [hunk 3]
```diff
@@ -213,13 +213,13 @@ class DataVariantValue : public DataValueWrapper
 
    DataValueWrapper* copy() const
    {
-      DataValueWrapper* pData = new DataVariantValue(&(this->DISAMBIGUATE_TEMPLATE mValue));
+      DataValueWrapper* pData = new DataVariantValue(&(this->mValue)); // T is not a template, it's a typename or class
```
Comment (tclarke) [inline Code/application/Utilities/DataVariantFactoryImp.cpp:216]: This code was required when Solaris was a supported system because the STL was a bit older and had some issues requiring this. Removing it everywhere is fine now.

**Review by tclarke — COMMENTED** [review]
Approving and merging. I looked a little deeper at that change from false to 0 and I think there's a deeper issue there that needs to be fixed. I believe it should be (std::string(), false) instead but I'm not 100% on that yet so I'll move it along as-is.

**Comment by tclarke** [comment]
Also, it's safe to ignore build pipeline failures right now. I'm just getting it setup to build and test for windows so it's always failing.
